File: stack/leetcode_20.cpp
/**
 * ## 有效的括号
 * 给定一个只包括 '('，')'，'{'，'}'，'['，']' 的字符串，判断字符串是否有效。
 * *** Notice ***
 * 1.左括号必须用相同类型的右括号闭合。
 * 2.左括号必须以正确的顺序闭合
 * 3.空字符串可被认为是有效字符串
 */

#include <iostream>
#include <stack>
#include <string>

using namespace std;

class Solution {
 public:
  bool isValid(string s);
};

bool Solution::isValid(string s) {
  stack<char> st;
  for (auto ch : s) {
    if (ch == '(' || ch == '[' || ch == '{') {
      st.push(ch);
    } else if (ch == ')' || ch == ']' || ch == '}') {
      if (st.empty()) return false;
      char stTop = st.top();
      if (stTop == '(' && ch == ')' || stTop == '[' && ch == ']' ||
          stTop == '{' && ch == '}') {
        st.pop();
      } else {
        return false;
      }
    } else {
      return false;
    }
  }
  return st.empty();
}

int main() {
  string testStr = "(){}";
  Solution solution;
  solution.isValid(testStr);
  
  return 0;
}

File: stack/leetcode_20.js
// leetcode20,括号匹配
// 执行用时：72 ms, 在所有 JavaScript 提交中击败了98.35%的用户
// 内存消耗：37.8 MB, 在所有 JavaScript 提交中击败了42.33%的用户
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
    let n = s.length
    if (n % 2 !== 0) return false

    let stack = []
    let map = new Map([
        [')', '('],
        [']', '['],
        ['}', '{']
    ])
    for (let i = 0; i < n; i++) {
        let char = s.charAt(i)
        if (map.has(char)) {
            if (stack.length === 0 || stack[stack.length - 1] !== map.get(char)) {
                return false
            }
            stack.pop()
        } else {
            stack.push(char)
        }
    }
    return stack.length === 0
};


let test = "([}}])"
console.log(isValid(test))